README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

# Aura UI

🚧 Warning! Work in progress. 🚧

There are things that may not work. API not stable and subject to breaking changes.

# Getting started

## Installation
Open a terminal and run the following command:

```bash
$ flutter pub add aura_ui
```

## Usage
Wrap your app with the `AuraUI` widget and set the light and dark themes.

```dart
import 'package:flutter/material.dart';
import 'package:aura_ui/aura_ui.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  static final _lightTheme = ATheme(
    primary: Colors.blue,
    secondary: Colors.blueGrey[200],
    success: Colors.green,
    warning: Colors.yellow,
    error: Colors.red,
    info: Colors.blue,
    background: Colors.white,
    surface100: Colors.white,
    surface200: Colors.grey[100],
    textPrimary: Colors.black,
    textSecondary: Colors.grey,
    divider: Colors.grey[300],
    border: Colors.grey[300],
  );

  static final _darkTheme = ATheme(
    primary: Colors.blue,
    secondary: Colors.blueGrey[700],
    success: Colors.green,
    warning: Colors.yellow,
    error: Colors.red,
    info: Colors.blue,
    background: Colors.black,
    surface100: Colors.grey[900],
    surface200: Colors.grey[800],
    textPrimary: Colors.white,
    textSecondary: Colors.grey,
    divider: Colors.grey[700],
    border: Colors.grey[700],
  );

  @override
  Widget build(BuildContext context) {
    return AuraUI(
      lightTheme: _lightTheme,
      darkTheme: _darkTheme,
      child: MaterialApp(
        title: 'Example',
        home: const Home(),
      ),
    );
  }
}
```
Now you can use AuraUI typography and colors in your widgets and switch themes easily.

```dart
import 'package:flutter/material.dart';

class Home extends StatelessWidget {
  const Home({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: AColor.background,
      body: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          const ThemeSettings(),
          AText.displayXL("Display XL").styles(
            color: AColor.secondary,
            fontWeight: AFontWeight.bold,
          ),
          AText.displayXL("Display XL 2").styles(
            color: AColor.error,
            fontFamily: AFontFamily.inter,
          ),
        ],
      ),
    );
  }
}

class ThemeSettings extends StatelessWidget {
  const ThemeSettings({super.key});

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        AText.displayXL("Theme Settings").styles(
          color: AColor.textPrimary,
        ),
        Row(
          children: [
            ElevatedButton(
              child: AText("Light Theme"),
              onPressed: () {
                AuraUI.setLightTheme();
              },
            ),
            ElevatedButton(
              child: AText("Dark Theme"),
              onPressed: () {
                AuraUI.setDarkTheme();
              },
            ),
          ],
        ),
      ],
    );
  }
}
```

# License

MIT License

Copyright (c) 2023 Raúl Colino

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

CHANGELOG.md
## 0.1.0

* First testable alpha release. 
* Typography and Colors added but not yet fully implemented.

## 0.0.1

* Initial alpha release.


lib/src/theme/aura_ui.dart
import 'package:flutter/material.dart';
import '../colors/a_color.dart';
import '../colors/a_theme.dart';

//Global variables to store the light and dark themes
ATheme? _lightTheme;
ATheme? _darkTheme;

enum ThemeType { light, dark }

class AuraUI extends StatefulWidget {
  //Properties
  //The child widget that will be wrapped by AuraUI. When the theme changes, the child widget will be rebuilt.
  final Widget child;
  //Establishes a theme selection and persists it in memory across hot reloads
  static ThemeType? _currentThemeType;
  //This is a global key that let us access the _AuraUIState from the AuraUI class
  //See: https://stackoverflow.com/a/60513911
  static final GlobalKey<_AuraUIState> _auraUIStateGlobalKey = GlobalKey();

  //Constructor
  AuraUI({
    Key? key,
    required ATheme lightTheme,
    required ATheme darkTheme,
    required this.child,
  }) : super(key: _auraUIStateGlobalKey) {
    //If theme not set, set the light theme
    if (_currentThemeType == null) {
      _lightTheme = lightTheme;
      _darkTheme = darkTheme;
      //Set light theme as default
      _currentThemeType = ThemeType.light;
      AColor.setTheme(lightTheme);
    }
  }

  //Methods
  static void _setThemeAndUpdateState(ATheme? theme) {
    if (theme == null) throw Exception("AuraUI ERROR: AuraUI widget not created. Please wrap your app with the AuraUI widget. Check the docs for more info.");
    AColor.setTheme(theme);
    try {
      //To use this, AuraUI must be a StatefulWidget
      _auraUIStateGlobalKey.currentState!
          .setState(() {}); // ignore: invalid_use_of_protected_member
    } catch (e) {
      //The current state (_auraUIKey.currentState) is null if there is no widget in the tree that
      // matches the global key: _auraUIStateGlobalKey, which means that the AuraUI widget is not in the widget tree.
      throw Exception("AuraUI ERROR: AuraUI widget not found in the widget tree.");
    }
  }

  static void setLightTheme() {
    if (_currentThemeType != ThemeType.light) {
      print("setLightTheme");
      _currentThemeType = ThemeType.light;
      _setThemeAndUpdateState(_lightTheme);
    }
  }

  static void setDarkTheme() {
    if (_currentThemeType != ThemeType.dark) {
      print("setDarkTheme");
      _currentThemeType = ThemeType.dark;
      _setThemeAndUpdateState(_darkTheme);
    }
  }

  @override
  State<AuraUI> createState() => _AuraUIState();
}

class _AuraUIState extends State<AuraUI> {
  Key refreshKey = UniqueKey();

  //State refresh based on: https://stackoverflow.com/a/50116077 and https://stackoverflow.com/a/73129922/463029
  void _refreshState() {
    setState(() {
      refreshKey = UniqueKey();
    });
  }

  @override
  Widget build(BuildContext context) {
    //print("AuraUI build");
    //Rebuild all the child widgets of AuraUI
    _refreshState();
    //Return the child widget wrapped in a KeyedSubtree widget to rebuild its children after the refreshKey changes. See: https://stackoverflow.com/a/50116077
    return KeyedSubtree(
      key: refreshKey,
      child: widget.child,
    );
  }
}


lib/src/colors/a_color.dart
// ignore_for_file: curly_braces_in_flow_control_structures

import 'dart:ui';

import 'a_theme.dart';

//You can't exclude this file from the dart formatter because the Flutter team has not implemented this feature yet -_-

class AColor {
  //Exceptions
  static final Exception _auraWidgetException = Exception("AuraUI ERROR: AuraUI widget not created. Please wrap your app with the AuraUI widget. Check the docs for more info.");
  static final Exception _colorNullException = Exception("AuraUI ERROR: Color cannot be null.");

  //Aura semantic colors
  static Color? _primary;
  static Color? _secondary;
  static Color? _success;
  static Color? _warning;
  static Color? _error;
  static Color? _info;
  static Color? _background;
  static Color? _surface100;
  static Color? _surface200;
  static Color? _textPrimary;
  static Color? _textSecondary;
  static Color? _divider;
  static Color? _border;

  //Getters
  static Color get primary { if (_primary == null) throw _auraWidgetException; else return _primary!; }
  static Color get secondary { if (_secondary == null) throw _auraWidgetException; else return _secondary!; }
  static Color get success { if (_success == null) throw _auraWidgetException; else return _success!; }
  static Color get warning { if (_warning == null) throw _auraWidgetException; else return _warning!; }
  static Color get error { if (_error == null) throw _auraWidgetException; else return _error!; }
  static Color get info { if (_info == null) throw _auraWidgetException; else return _info!; }
  static Color get background { if (_background == null) throw _auraWidgetException; else return _background!; }
  static Color get surface100 { if (_surface100 == null) throw _auraWidgetException; else return _surface100!; }
  static Color get surface200 { if (_surface200 == null) throw _auraWidgetException; else return _surface200!; }
  static Color get textPrimary { if (_textPrimary == null) throw _auraWidgetException; else return _textPrimary!; }
  static Color get textSecondary { if (_textSecondary == null) throw _auraWidgetException; else return _textSecondary!; }
  static Color get divider { if (_divider == null) throw _auraWidgetException; else return _divider!; }
  static Color get border { if (_border == null) throw _auraWidgetException; else return _border!; }

  //Setters
  static set primary(Color? color) { if (color == null) throw _colorNullException; else _primary = color; }
  static set secondary(Color? color) { if (color == null) throw _colorNullException; else _secondary = color; }
  static set success(Color? color) { if (color == null) throw _colorNullException; else _success = color; }
  static set warning(Color? color) { if (color == null) throw _colorNullException; else _warning = color; }
  static set error(Color? color) { if (color == null) throw _colorNullException; else _error = color; }
  static set info(Color? color) { if (color == null) throw _colorNullException; else _info = color; }
  static set background(Color? color) { if (color == null) throw _colorNullException; else _background = color; }
  static set surface100(Color? color) { if (color == null) throw _colorNullException; else _surface100 = color; }
  static set surface200(Color? color) { if (color == null) throw _colorNullException; else _surface200 = color; }
  static set textPrimary(Color? color) { if (color == null) throw _colorNullException; else _textPrimary = color; }
  static set textSecondary(Color? color) { if (color == null) throw _colorNullException; else _textSecondary = color; }
  static set divider(Color? color) { if (color == null) throw _colorNullException; else _divider = color; }
  static set border(Color? color) { if (color == null) throw _colorNullException; else _border = color; }

  //Constructor
  AColor._();

  static void setTheme(ATheme theme) {
    primary = theme.primary!;
    secondary = theme.secondary!;
    success = theme.success!;
    warning = theme.warning!;
    error = theme.error!;
    info = theme.info!;
    background = theme.background!;
    surface100 = theme.surface100!;
    surface200 = theme.surface200!;
    textPrimary = theme.textPrimary!;
    textSecondary = theme.textSecondary!;
    divider = theme.divider!;
    border = theme.border!;
  }
}

// class AColor {
  
//   static final Exception _exception = Exception("AuraUI ERROR: AuraUI widget not created. Please wrap your app with the AuraUI widget. Check the docs for more info.");

//   //Aura semantic colors
//   static AColorProvider? _primary;
//   static AColorProvider? _secondary;
//   static AColorProvider? _success;
//   static AColorProvider? _warning;
//   static AColorProvider? _error;
//   static AColorProvider? _info;
//   static AColorProvider? _background;
//   static AColorProvider? _surface100;
//   static AColorProvider? _surface200;
//   static AColorProvider? _textPrimary;
//   static AColorProvider? _textSecondary;
//   static AColorProvider? _divider;
//   static AColorProvider? _border;

//   static AColorProvider get primary {
//     if (_primary == null) throw _exception;
//     return _primary!;
//   }
//   static AColorProvider get secondary {
//     if (_secondary == null) throw _exception;
//     return _secondary!;
//   }
//   static AColorProvider get success {
//     if (_success == null) throw _exception;
//     return _success!;
//   }
//   static AColorProvider get warning {
//     if (_warning == null) throw _exception;
//     return _warning!;
//   }
//   static AColorProvider get error {
//     if (_error == null) throw _exception;
//     return _error!;
//   }
//   static AColorProvider get info {
//     if (_info == null) throw _exception;
//     return _info!;
//   }
//   static AColorProvider get background {
//     if (_background == null) throw _exception;
//     return _background!;
//   }
//   static AColorProvider get surface100 {
//     if (_surface100 == null) throw _exception;
//     return _surface100!;
//   }
//   static AColorProvider get surface200 {
//     if (_surface200 == null) throw _exception;
//     return _surface200!;
//   }
//   static AColorProvider get textPrimary {
//     if (_textPrimary == null) throw _exception;
//     return _textPrimary!;
//   }
//   static AColorProvider get textSecondary {
//     if (_textSecondary == null) throw _exception;
//     return _textSecondary!;
//   }
//   static AColorProvider get divider {
//     if (_divider == null) throw _exception;
//     return _divider!;
//   }
//   static AColorProvider get border {
//     if (_border == null) throw _exception;
//     return _border!;
//   }


//   static set primary(Color? color) {
//     if (color == null) throw Exception("AuraUI ERROR: color cannot be null.");
//     _primary = AColorProvider(color);
//   }
//   static set secondary(Color? color) {
//     if (color == null) throw Exception("AuraUI ERROR: color cannot be null.");
//     _secondary = AColorProvider(color);
//   }
//   static set success(Color? color) {
//     if (color == null) throw Exception("AuraUI ERROR: color cannot be null.");
//     _success = AColorProvider(color);
//   }
//   static set warning(Color? color) {
//     if (color == null) throw Exception("AuraUI ERROR: color cannot be null.");
//     _warning = AColorProvider(color);
//   }
//   static set error(Color? color) {
//     if (color == null) throw Exception("AuraUI ERROR: color cannot be null.");
//     _error = AColorProvider(color);
//   }
//   static set info(Color? color) {
//     if (color == null) throw Exception("AuraUI ERROR: color cannot be null.");
//     _info = AColorProvider(color);
//   }
//   static set background(Color? color) {
//     if (color == null) throw Exception("AuraUI ERROR: color cannot be null.");
//     _background = AColorProvider(color);
//   }
//   static set surface100(Color? color) {
//     if (color == null) throw Exception("AuraUI ERROR: color cannot be null.");
//     _surface100 = AColorProvider(color);
//   }
//   static set surface200(Color? color) {
//     if (color == null) throw Exception("AuraUI ERROR: color cannot be null.");
//     _surface200 = AColorProvider(color);
//   }
//   static set textPrimary(Color? color) {
//     if (color == null) throw Exception("AuraUI ERROR: color cannot be null.");
//     _textPrimary = AColorProvider(color);
//   }
//   static set textSecondary(Color? color) {
//     if (color == null) throw Exception("AuraUI ERROR: color cannot be null.");
//     _textSecondary = AColorProvider(color);
//   }
//   static set divider(Color? color) {
//     if (color == null) throw Exception("AuraUI ERROR: color cannot be null.");
//     _divider = AColorProvider(color);
//   }
//   static set border(Color? color) {
//     if (color == null) throw Exception("AuraUI ERROR: color cannot be null.");
//     _border = AColorProvider(color);
//   }

//   //Constructor
//   AColor._();

//   static void setTheme(ATheme theme) {
//     primary = AColorProvider(theme.primary!);
//     secondary = AColorProvider(theme.secondary!);
//     success = AColorProvider(theme.success!);
//     warning = AColorProvider(theme.warning!);
//     error = AColorProvider(theme.error!);
//     info = AColorProvider(theme.info!);
//     background = AColorProvider(theme.background!);
//     surface100 = AColorProvider(theme.surface100!);
//     surface200 = AColorProvider(theme.surface200!);
//     textPrimary = AColorProvider(theme.textPrimary!);
//     textSecondary = AColorProvider(theme.textSecondary!);
//     divider = AColorProvider(theme.divider!);
//     border = AColorProvider(theme.border!);
//   }
// }



// class AColorProvider extends Color {
//   final Color? _color;
//   AColorProvider(this._color) : super(_color?.value ?? 0xFF00FF00);

//   Color get color => Color(value);

//   @override
//   int get value {
//     try {
//       return _color!.value;
//     } catch (e) {
//       throw Exception("AuraUI ERROR: AuraUI widget not created. Please wrap your app with the AuraUI widget. Check the docs for more info.");
//     }
//   }
// }



lib/src/typography/a_text.dart
import 'dart:ui';

import 'package:flutter/widgets.dart';

import 'a_font_family.dart';
import 'a_font_weight.dart';

///Aura Text widget
class AText extends Text {
  //Exclusive attributes of the class
  final AFontWeight? fontWeight;
  final AFontFamily? fontFamily;
  final String? package; //Added because TextStyle package is private and doesn't have a getter

  //AuraUI text types
  //Display XL
  static AText displayXL(String data) => AText._(data).styles(
        fontSize: 96,
        fontWeight: AFontWeight.bold,
        letterSpacing: -1.5,
      );

  //Display L
  static AText displayL(String data) => AText._(data).styles(
        fontSize: 60,
        fontWeight: AFontWeight.bold,
        letterSpacing: -0.5,
      );

  //Body M
  static AText bodyM(String data) => AText._(data).styles(
        fontSize: 16,
        fontWeight: AFontWeight.regular,
        letterSpacing: 0.5,
      );

  //Constructor
  AText._(
    String data, {
    Key? key,
    StrutStyle? strutStyle,
    TextAlign? textAlign,
    TextDirection? textDirection,
    Locale? locale,
    bool? softWrap,
    TextOverflow? overflow,
    double? textScaleFactor,
    int? maxLines,
    String? semanticsLabel,
    TextWidthBasis? textWidthBasis,
    TextHeightBehavior? textHeightBehavior,
    Color? selectionColor,
    //'TextStyle style' attributes from standard Text widget
    bool? inherit,
    Color? color,
    Color? backgroundColor,
    double? fontSize = 14,
    this.fontWeight = AFontWeight.regular,
    FontStyle? fontStyle,
    double? letterSpacing = 0.0,
    double? wordSpacing,
    TextBaseline? textBaseline,
    double? height = 1.0,
    Paint? foreground,
    Paint? background,
    List<Shadow>? shadows,
    List<FontFeature>? fontFeatures,
    TextDecoration? decoration,
    Color? decorationColor,
    TextDecorationStyle? decorationStyle,
    double? decorationThickness,
    String? debugLabel,
    this.fontFamily = AFontFamily.inter,
    List<String>? fontFamilyFallback,
    this.package = "theme_kit",
  }) : super(
          data,
          key: key,
          style: TextStyle(
            inherit: inherit ?? true,
            color: color,
            backgroundColor: backgroundColor,
            fontSize: fontSize,
            fontWeight: fontWeight?.value,
            fontStyle: fontStyle,
            letterSpacing: letterSpacing,
            wordSpacing: wordSpacing,
            textBaseline: textBaseline,
            height: height,
            locale: locale,
            foreground: foreground,
            background: background,
            shadows: shadows,
            fontFeatures: fontFeatures,
            decoration: decoration,
            decorationColor: decorationColor,
            decorationStyle: decorationStyle,
            decorationThickness: decorationThickness,
            debugLabel: debugLabel,
            fontFamily: fontFamily?.name,
            fontFamilyFallback: fontFamilyFallback,
            package: package,
          ),
          strutStyle: strutStyle,
          textAlign: textAlign,
          textDirection: textDirection,
          locale: locale,
          softWrap: softWrap,
          overflow: overflow,
          textScaleFactor: textScaleFactor,
          maxLines: maxLines,
          semanticsLabel: semanticsLabel,
          textWidthBasis: textWidthBasis,
          textHeightBehavior: textHeightBehavior,
          selectionColor: selectionColor,
        );

  //CopyWith
  AText _copyWith({
    String? data,
    Key? key,
    StrutStyle? strutStyle,
    TextAlign? textAlign,
    TextDirection? textDirection,
    Locale? locale,
    bool? softWrap,
    TextOverflow? overflow,
    double? textScaleFactor,
    int? maxLines,
    String? semanticsLabel,
    TextWidthBasis? textWidthBasis,
    TextHeightBehavior? textHeightBehavior,
    Color? selectionColor,
    //'TextStyle style' attributes from standard Text widget
    bool? inherit,
    Color? color,
    Color? backgroundColor,
    double? fontSize,
    AFontWeight? fontWeight,
    FontStyle? fontStyle,
    double? letterSpacing,
    double? wordSpacing,
    TextBaseline? textBaseline,
    double? height,
    Paint? foreground,
    Paint? background,
    List<Shadow>? shadows,
    List<FontFeature>? fontFeatures,
    TextDecoration? decoration,
    Color? decorationColor,
    TextDecorationStyle? decorationStyle,
    double? decorationThickness,
    String? debugLabel,
    AFontFamily? fontFamily,
    List<String>? fontFamilyFallback,
    String? package,
  }) {
    return AText._(
      data ?? this.data ?? "",
      key: key ?? this.key,
      strutStyle: strutStyle ?? this.strutStyle,
      textAlign: textAlign ?? this.textAlign,
      textDirection: textDirection ?? this.textDirection,
      locale: locale ?? this.locale,
      softWrap: softWrap ?? this.softWrap,
      overflow: overflow ?? this.overflow,
      textScaleFactor: textScaleFactor ?? this.textScaleFactor,
      maxLines: maxLines ?? this.maxLines,
      semanticsLabel: semanticsLabel ?? this.semanticsLabel,
      textWidthBasis: textWidthBasis ?? this.textWidthBasis,
      textHeightBehavior: textHeightBehavior ?? this.textHeightBehavior,
      selectionColor: selectionColor ?? this.selectionColor,
      //'TextStyle style' attributes from standard Text widget
      inherit: inherit ?? style?.inherit,
      color: color ?? style?.color,
      backgroundColor: backgroundColor ?? style?.backgroundColor,
      fontSize: fontSize ?? style?.fontSize,
      fontWeight: fontWeight ?? this.fontWeight,
      fontStyle: fontStyle ?? style?.fontStyle,
      letterSpacing: letterSpacing ?? style?.letterSpacing,
      wordSpacing: wordSpacing ?? style?.wordSpacing,
      textBaseline: textBaseline ?? style?.textBaseline,
      height: height ?? style?.height,
      foreground: foreground ?? style?.foreground,
      background: background ?? style?.background,
      shadows: shadows ?? style?.shadows,
      fontFeatures: fontFeatures ?? style?.fontFeatures,
      decoration: decoration ?? style?.decoration,
      decorationColor: decorationColor ?? style?.decorationColor,
      decorationStyle: decorationStyle ?? style?.decorationStyle,
      decorationThickness: decorationThickness ?? style?.decorationThickness,
      debugLabel: debugLabel ?? style?.debugLabel,
      fontFamily: fontFamily ?? this.fontFamily,
      fontFamilyFallback: fontFamilyFallback ?? style?.fontFamilyFallback,
      package: package ?? this.package,
    );
  }

  AText styles({
    Key? key,
    StrutStyle? strutStyle,
    TextAlign? textAlign,
    TextDirection? textDirection,
    Locale? locale,
    bool? softWrap,
    TextOverflow? overflow,
    double? textScaleFactor,
    int? maxLines,
    String? semanticsLabel,
    TextWidthBasis? textWidthBasis,
    TextHeightBehavior? textHeightBehavior,
    Color? selectionColor,
    //'TextStyle style' attributes from standard Text widget
    bool? inherit,
    Color? color,
    Color? backgroundColor,
    double? fontSize,
    AFontWeight? fontWeight,
    FontStyle? fontStyle,
    double? letterSpacing,
    double? wordSpacing,
    TextBaseline? textBaseline,
    double? height,
    Paint? foreground,
    Paint? background,
    List<Shadow>? shadows,
    List<FontFeature>? fontFeatures,
    TextDecoration? decoration,
    Color? decorationColor,
    TextDecorationStyle? decorationStyle,
    double? decorationThickness,
    String? debugLabel,
    AFontFamily? fontFamily,
    List<String>? fontFamilyFallback,
    String? package,
  }) =>
      _copyWith(
        key: key,
        strutStyle: strutStyle,
        textAlign: textAlign,
        textDirection: textDirection,
        locale: locale,
        softWrap: softWrap,
        overflow: overflow,
        textScaleFactor: textScaleFactor,
        maxLines: maxLines,
        semanticsLabel: semanticsLabel,
        textWidthBasis: textWidthBasis,
        textHeightBehavior: textHeightBehavior,
        selectionColor: selectionColor,
        inherit: inherit,
        color: color,
        backgroundColor: backgroundColor,
        fontSize: fontSize,
        fontWeight: fontWeight,
        fontStyle: fontStyle,
        letterSpacing: letterSpacing,
        wordSpacing: wordSpacing,
        textBaseline: textBaseline,
        height: height,
        foreground: foreground,
        background: background,
        shadows: shadows,
        fontFeatures: fontFeatures,
        decoration: decoration,
        decorationColor: decorationColor,
        decorationStyle: decorationStyle,
        decorationThickness: decorationThickness,
        debugLabel: debugLabel,
        fontFamily: fontFamily,
        fontFamilyFallback: fontFamilyFallback,
        package: package,
      );

//Alternative version:
  // AText.displayXL(
  //   String data, {
  //   Key? key,
  //   StrutStyle? strutStyle,
  //   TextAlign? textAlign,
  //   TextDirection? textDirection,
  //   Locale? locale,
  //   bool? softWrap,
  //   TextOverflow? overflow,
  //   double? textScaleFactor,
  //   int? maxLines,
  //   String? semanticsLabel,
  //   TextWidthBasis? textWidthBasis,
  //   TextHeightBehavior? textHeightBehavior,
  //   Color? selectionColor,
  //   //'TextStyle style' attributes from standard Text widget
  //   bool? inherit,
  //   Color? color,
  //   Color? backgroundColor,
  //   double? fontSize = 72,
  //   this.fontWeight = AFontWeight.bold,
  //   FontStyle? fontStyle,
  //   double? letterSpacing = -1.44,
  //   double? wordSpacing,
  //   TextBaseline? textBaseline,
  //   double? height = 1.08333,
  //   Paint? foreground,
  //   Paint? background,
  //   List<Shadow>? shadows,
  //   List<FontFeature>? fontFeatures,
  //   TextDecoration? decoration,
  //   Color? decorationColor,
  //   TextDecorationStyle? decorationStyle,
  //   double? decorationThickness,
  //   String? debugLabel,
  //   this.fontFamily = AFontFamily.inter,
  //   List<String>? fontFamilyFallback,
  //   this.package = "aura_ui",
  // }) : super(
  //         data,
  //         key: key,
  //         style: TextStyle(
  //           inherit: inherit ?? true,
  //           color: color,
  //           backgroundColor: backgroundColor,
  //           fontSize: fontSize,
  //           fontWeight: fontWeight?.value,
  //           fontStyle: fontStyle,
  //           letterSpacing: letterSpacing,
  //           wordSpacing: wordSpacing,
  //           textBaseline: textBaseline,
  //           height: height,
  //           locale: locale,
  //           foreground: foreground,
  //           background: background,
  //           shadows: shadows,
  //           fontFeatures: fontFeatures,
  //           decoration: decoration,
  //           decorationColor: decorationColor,
  //           decorationStyle: decorationStyle,
  //           decorationThickness: decorationThickness,
  //           debugLabel: debugLabel,
  //           fontFamily: fontFamily?.name,
  //           fontFamilyFallback: fontFamilyFallback,
  //           package: package,
  //         ),
  //         strutStyle: strutStyle,
  //         textAlign: textAlign,
  //         textDirection: textDirection,
  //         locale: locale,
  //         softWrap: softWrap,
  //         overflow: overflow,
  //         textScaleFactor: textScaleFactor,
  //         maxLines: maxLines,
  //         semanticsLabel: semanticsLabel,
  //         textWidthBasis: textWidthBasis,
  //         textHeightBehavior: textHeightBehavior,
  //         selectionColor: selectionColor,
  //       );
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:aura_ui/aura_ui.dart';


void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  static final _lightTheme = ATheme(
    primary: Colors.blue,
    secondary: Colors.blueGrey[200],
    success: Colors.green,
    warning: Colors.yellow,
    error: Colors.red,
    info: Colors.blue,
    background: Colors.white,
    surface100: Colors.white,
    surface200: Colors.grey[100],
    textPrimary: Colors.black,
    textSecondary: Colors.grey,
    divider: Colors.grey[300],
    border: Colors.grey[300],
  );

  static final _darkTheme = ATheme(
    primary: Colors.blue,
    secondary: Colors.blueGrey[700],
    success: Colors.green,
    warning: Colors.yellow,
    error: Colors.red,
    info: Colors.blue,
    background: Colors.black,
    surface100: Colors.grey[900],
    surface200: Colors.grey[800],
    textPrimary: Colors.white,
    textSecondary: Colors.grey,
    divider: Colors.grey[700],
    border: Colors.grey[700],
  );

  @override
  Widget build(BuildContext context) {
    try {
      return AuraUI(
        lightTheme: _lightTheme,
        darkTheme: _darkTheme,
        child: MaterialApp(
          title: 'Example',
          theme: ThemeData(
            primarySwatch: Colors.blue,
          ),
          home: const Home(),
        ),
      );
    } catch (e) {
      print(e);
      return Container(
        child: const Text("error"),
      );
    }
  }
}

// // Erroneous example. This is the same as above, but without AuraUI. It will show error colors.
// class MyApp extends StatelessWidget {
//   const MyApp({super.key});
//   @override
//   Widget build(BuildContext context) {
//     try {
//       return MaterialApp(
//         title: 'Example',
//         theme: ThemeData(
//           primarySwatch: Colors.blue,
//         ),
//         home: const Home(),
//       );
//     } catch (e) {
//       print(e);
//       return Container(
//         child: const Text("error"),
//       );
//     }
//   }
// }

class Home extends StatelessWidget {
  const Home({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: AColor.background,
      body: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          const ThemeSettings(),
          AText.displayXL("Display XL").styles(
            color: AColor.secondary,
            fontWeight: AFontWeight.bold,
          ),
          AText.displayL("Display L").styles(
            color: AColor.error,
            fontFamily: AFontFamily.inter,
          ),
        ],
      ),
    );
  }
}

class ThemeSettings extends StatelessWidget {
  const ThemeSettings({super.key});

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        AText.displayXL("Theme Settings").styles(
          color: AColor.textPrimary,
        ),
        Row(
          children: [
            ElevatedButton(
              child: AText.bodyM("Light Theme"),
              onPressed: () {
                AuraUI.setLightTheme();
              },
            ),
            ElevatedButton(
              child: AText.bodyM("Dark Theme"),
              onPressed: () {
                AuraUI.setDarkTheme();
              },
            ),
          ],
        ),
      ],
    );
  }
}
